data/transactions/logic/eval.go

```diff
@@ -3183,25 +3273,205 @@ func opAppParamsGet(cx *evalContext) {
 	cx.stack = append(cx.stack, stackValue{Uint: exist})
 }
 
-func opLog(cx *evalContext) {
+func opLog(cx *EvalContext) {
 	last := len(cx.stack) - 1
 
-	if cx.logCalls == MaxLogCalls {
+	if len(cx.Logs) == MaxLogCalls {
 		cx.err = fmt.Errorf("too many log calls in program. up to %d is allowed", MaxLogCalls)
 		return
 	}
-	cx.logCalls++
 	log := cx.stack[last]
 	cx.logSize += len(log.Bytes)
 	if cx.logSize > MaxLogSize {
 		cx.err = fmt.Errorf("program logs too large. %d bytes >  %d bytes limit", cx.logSize, MaxLogSize)
 		return
 	}
-	// write log to applyData
-	err := cx.Ledger.AppendLog(&cx.Txn.Txn, string(log.Bytes))
+	cx.Logs = append(cx.Logs, string(log.Bytes))
+	cx.stack = cx.stack[:last]
+}
+
+func authorizedSender(cx *EvalContext, addr basics.Address) bool {
+	appAddr, err := cx.getApplicationAddress()
+	if err != nil {
+		return false
+	}
+	return appAddr == cx.Ledger.Authorizer(addr)
+}
+
+func opTxBegin(cx *EvalContext) {
+	if cx.subtxn != nil {
+		cx.err = errors.New("tx_begin without tx_submit")
```

Comment:
No nesting allowed.  We only have one place to store txn data, so tx_begin cleans out the subtxn field. (This will have to change if we want inner groups, not just inner txns, but even then, no nesting.)